src/core/net/tcp6.cpp

```diff
@@ -302,27 +456,222 @@ Error Tcp::Listener::Deinitialize(void)
+     */
+    uint32_t header[15];
+
+    uint16_t length = aIp6Header.GetPayloadLength();
+    uint8_t  headerSize;
+
+    struct ip6_hdr *ip6Header;
+    struct tcphdr * tcpHeader;
+
+    Endpoint *endpoint;
+    Endpoint *endpointPrev;
+
+    Listener *listener;
+    Listener *listenerPrev;
+
+    VerifyOrExit(length == aMessage.GetLength() - aMessage.GetOffset(), error = kErrorParse);
+    VerifyOrExit(length >= sizeof(Tcp::Header), error = kErrorParse);
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset() + offsetof(struct tcphdr, th_off_x2), &headerSize, 1));
+    headerSize = static_cast<uint8_t>((headerSize >> TH_OFF_SHIFT) << 2);
+    VerifyOrExit(headerSize >= sizeof(struct tcphdr) && headerSize <= sizeof(header) &&
+                     static_cast<uint16_t>(headerSize) <= length,
+                 error = kErrorParse);
+    SuccessOrExit(error = Checksum::VerifyMessageChecksum(aMessage, aMessageInfo, kProtoTcp));
+    SuccessOrExit(error = aMessage.Read(aMessage.GetOffset(), &header[0], headerSize));
+
+    ip6Header = reinterpret_cast<struct ip6_hdr *>(&aIp6Header);
+    tcpHeader = reinterpret_cast<struct tcphdr *>(&header[0]);
+    tcp_fields_to_host(tcpHeader);
+
+    aMessageInfo.mPeerPort = HostSwap16(tcpHeader->th_sport);
+    aMessageInfo.mSockPort = HostSwap16(tcpHeader->th_dport);
+
+    endpoint = mEndpoints.FindMatching(aMessageInfo, endpointPrev);
+    if (endpoint != nullptr)
     {
-        OT_UNUSED_VARIABLE(active);
+        struct signals  sig;
+        otLinkedBuffer *priorHead = lbuf_head(&endpoint->mTcb.sendbuf);
+        memset(&sig, 0x00, sizeof(sig));
+        int nextAction = tcp_input(ip6Header, tcpHeader, &aMessage, &endpoint->mTcb, nullptr, &sig);
+        if (nextAction != RELOOKUP_REQUIRED)
+        {
+            if (endpoint->mTcb.t_state != TCP6S_CLOSED)
+            {
+                ProcessSignals(*endpoint, priorHead, sig);
+            }
+            ExitNow();
+        }
+        /* If the matching socket was in the TIME-WAIT state, then we try passive sockets. */
     }
 
-    for (Listener &passive : mListeners)
+    listener = mListeners.FindMatching(aMessageInfo, listenerPrev);
+    if (listener != nullptr)
     {
-        OT_UNUSED_VARIABLE(passive);
+        tcp_input(ip6Header, tcpHeader, &aMessage, nullptr, &listener->mTcbListen, nullptr);
+        ExitNow();
     }
 
+    tcp_dropwithreset(ip6Header, tcpHeader, nullptr, &InstanceLocator::GetInstance(), length - headerSize,
+                      ECONNREFUSED);
+
+exit:
     return error;
 }
 
+void Tcp::ProcessSignals(Endpoint &aEndpoint, otLinkedBuffer *aPriorHead, struct signals &aSignals)
+{
+    if (aEndpoint.mSendDoneCallback != nullptr)
+    {
+        otLinkedBuffer *curr = aPriorHead;
+        for (int i = 0; i != aSignals.links_popped; i++)
+        {
+            otLinkedBuffer *next = curr->mNext;
+            curr->mNext          = nullptr;
+            aEndpoint.mSendDoneCallback(&aEndpoint, curr);
```

Comment:
We are calling user's callback here. Whenever we invoke a callback we need to consider that the user may call other functions/method to change stuff before we return from the callback. Is this possible here? Anything we need to check/validate after returning from callback.

In general, I would suggest to try to invoke the callback as the last statement (but I know there are situations when it may not be possible).